### ISSUE TITLE
Ringtone picker crashes on Q

### ISSUE THREAD

**Issue by dirkam** [consp1racy/android-support-preference#120]
Can you please take a look at this? It crashes on devices running Q beta (latest beta 6 and possibly beta 5).

```java
Fatal Exception: java.lang.NumberFormatException: For input string: "default"
       at java.lang.Long.parseLong + 594(Long.java:594)
       at java.lang.Long.parseLong + 636(Long.java:636)
       at android.content.ContentUris.parseId + 89(ContentUris.java:89)
       at android.media.RingtoneManager.getRingtonePosition + 513(RingtoneManager.java:513)
       at net.xpece.android.support.preference.XpRingtonePreferenceDialogFragment.onPrepareDialogBuilder + 310(XpRingtonePreferenceDialogFragment.java:310)
       at androidx.preference.PreferenceDialogFragmentCompat.onCreateDialog + 155(PreferenceDialogFragmentCompat.java:155)
       at net.xpece.android.support.preference.XpRingtonePreferenceDialogFragment.onCreateDialog + 179(XpRingtonePreferenceDialogFragment.java:179)
       at androidx.fragment.app.DialogFragment.onGetLayoutInflater + 330(DialogFragment.java:330)
       at net.xpece.android.support.preference.XpRingtonePreferenceDialogFragment.onActivityCreated + 169(XpRingtonePreferenceDialogFragment.java:169)
       at androidx.fragment.app.Fragment.performActivityCreated + 2460(Fragment.java:2460)
       at androidx.fragment.app.FragmentManagerImpl.moveToState + 1483(FragmentManagerImpl.java:1483)
       at androidx.fragment.app.FragmentManagerImpl.moveFragmentToExpectedState + 1784(FragmentManagerImpl.java:1784)
       at androidx.fragment.app.FragmentManagerImpl.moveToState + 1852(FragmentManagerImpl.java:1852)
       at androidx.fragment.app.BackStackRecord.executeOps + 802(BackStackRecord.java:802)
       at androidx.fragment.app.FragmentManagerImpl.executeOps + 2625(FragmentManagerImpl.java:2625)
       at androidx.fragment.app.FragmentManagerImpl.executeOpsTogether + 2411(FragmentManagerImpl.java:2411)
       at androidx.fragment.app.FragmentManagerImpl.removeRedundantOperationsAndExecute + 2366(FragmentManagerImpl.java:2366)
       at androidx.fragment.app.FragmentManagerImpl.execPendingActions + 2273(FragmentManagerImpl.java:2273)
       at androidx.fragment.app.FragmentManagerImpl$1.run + 733(FragmentManagerImpl.java:733)
       at android.os.Handler.handleCallback + 883(Handler.java:883)
       at android.os.Handler.dispatchMessage + 100(Handler.java:100)
       at android.os.Looper.loop + 214(Looper.java:214)
       at android.app.ActivityThread.main + 7356(ActivityThread.java:7356)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run + 492(RuntimeInit.java:492)
       at com.android.internal.os.ZygoteInit.main + 930(ZygoteInit.java:930)
```

**Comment by consp1racy** [comment]
Until Android Q is released to AOSP I'm blind.

I can implement a fallback to system picker in the library. In the meantime the easiest workaround for you is to put the following in your settings fragment:

    @Override
    public void onDisplayPreferenceDialog(@NonNull Preference preference) {
        if (preference instanceof RingtonePreference && Build.VERSION.SDK_INT >= 29) {
            final Intent i = ((RingtonePreference) preference).buildRingtonePickerIntent();
            startActivityForResult(i, 0xf00);
        } else {
            super.onDisplayPreferenceDialog(preference);
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == 0xf00) {
            findRingtonePreference().onActivityResult(data);
        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    private RingtonePreference findRingtonePreference() {
        // get your ringtone preference by key here
    }

Let's keep the issue open for tracking.

**Comment by consp1racy** [comment]
Depending on your input and preference setup the workaround may not work for you. **The system picker may also crash.** Basically you can feed it
* Default `Uri` only if you enabled Default item in the list.
* Silent `Uri` only if you enabled Silent item in the list.
* Any other `Uri` that ends with a number.

Previously `RingtoneManager.getRingtonePosition` would return `-1`. Now it crashes when the `Uri` doesn't end with a number.

~I'll report this to Google.~ Reported to Google: https://issuetracker.google.com/issues/139935440

---

As for your `Uri` that's ending with `"default"`  I don't know where that came from. The standard `Uri`s are:
* `content://com.android.settings/system/ringtone`
* `content://com.android.settings/system/notification_sound`
* `content://com.android.settings/system/alarm_alert`

What can you tell me about the device the crash occurred on?

**Comment by consp1racy** [comment]
Already tracked here: https://issuetracker.google.com/issues/139146343

**Comment by consp1racy** [comment]
This is now fixed in 4dae39186b (2.x-dev) and 6d5ebc7934 (3.x-dev).

`RingtonePreference.buildRingtonePickerIntent` will also safeguard against this issue.

**Comment by consp1racy** [comment]
Released in 3.0.0 on Maven Central.
